Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,611 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2019 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import NIO
+import NIOConcurrencyHelpers
+import NIOHTTP1
+import NIOTLS
+
+/// A connection pool that manages and creates new connections to hosts respecting the specified preferences
+class ConnectionPool {
+    /// The configuration used to bootstrap new HTTP connections
+    private let configuration: HTTPClient.Configuration
+
+    /// The main data structure used by the `ConnectionPool` to retreive and create connections associated
+    /// to a a given `Key` .
+    /// - Warning: This property shouldn't be directly accessed, use the `ConnectionPool` subscript instead
+    var _connectionProviders: [Key: HTTP1ConnectionProvider] = [:]
+
+    /// The lock used by the connection pool used to ensure correct synchronization of accesses to `_connectionProviders`
+    ///
+    /// Accesses that only need synchronization for the span of a get or set can use the subscript instead of this property
+    ///
+    /// - Warning: This lock should always be acquired *before* `HTTP1ConnectionProvider`s `stateLock` if used in combination with it.
+    private let connectionProvidersLock = Lock()
+
+    init(configuration: HTTPClient.Configuration) {
+        self.configuration = configuration
+    }
+
+    /// This enables the connection pool to store and retreive `HTTP1ConnectionProvider`s
+    /// by ensuring thread safety using the `connectionProvidersLock`.
+    private subscript(key: Key) -> HTTP1ConnectionProvider? {
```

Comment:
@Lukasa Thanks, I notice `set` isn't used anywhere indeed, removed in 06fc889cb035b7d46d0eba922669948c9cd29b7c

Reply:
@Lukasa Actually, removed the `subscript` entirely in b610184a611ded0582c2df80d624f66f9fb0a4e4 because there were really few callers for it and I think it just made things more complicated